### PR TITLE
OpenSSL 1.1.1 support

### PR DESCRIPTION
**Pull request by sfackler**

### CODE REVIEW

**Comment by sfackler** [comment]
Tests are going to fail because of https://github.com/openssl/openssl/issues/5362.

**Comment by sfackler** [comment]
cc @Ralith

File: openssl-sys/build.rs [hunk 1]
```diff
@@ -343,8 +343,10 @@ RUST_LIBRESSL_251
 RUST_LIBRESSL_250
 #elif defined (LIBRESSL_VERSION_NUMBER) && LIBRESSL_VERSION_NUMBER < 0x20500000
 RUST_LIBRESSL_OLD
-#elif OPENSSL_VERSION_NUMBER >= 0x10101000
+#elif OPENSSL_VERSION_NUMBER >= 0x10102000
```
Comment (Ralith) [inline openssl-sys/build.rs]: Why reject 1.1.2, given the effort they're going to to maintain ABI compat?

File: openssl/src/ssl/test.rs [hunk 2]
```diff
@@ -1038,6 +1041,13 @@ fn tmp_dh_callback() {
 
     let stream = TcpStream::connect(("127.0.0.1", port)).unwrap();
     let mut ctx = SslContext::builder(SslMethod::tls()).unwrap();
+    // TLS 1.3 has no DH suites, and openssl isn't happy if the max version has no suites :(
```
Comment (Ralith) [inline openssl/src/ssl/test.rs]: Is this a temporary problem?

File: systest/build.rs [hunk 3]
```diff
@@ -29,12 +29,10 @@ fn main() {
...
-        (
-            env::var("DEP_OPENSSL_VERSION"),
-            env::var("DEP_OPENSSL_PATCH"),
-        )
-    {
+    if let (Ok(version), Ok(patch)) = (
+        env::var("DEP_OPENSSL_VERSION"),
+        env::var("DEP_OPENSSL_PATCH"),
+    ) {
         cfg.cfg(&format!("ossl{}{}", version, patch), None);
```
Comment (Ralith) [inline systest/build.rs]: Missing ossl111 implies ossl110 here?

File: openssl/src/ssl/test.rs (hunk 2 above)
Comment (sfackler) [reply openssl/src/ssl/test.rs]: Dunno

File: openssl-sys/build.rs (hunk 1 above)
Comment (sfackler) [reply openssl-sys/build.rs]: Their notion of ABI compat is not all that strict. Several constants have changed values between 1.1.0 and 1.1.1. 1.1.2 isn't going to be releasing for over a year, so I don't think this check is really going to be a problem for people.

File: openssl/src/ssl/test.rs (hunk 2 above)
Comment (sfackler) [reply openssl/src/ssl/test.rs]: I'm actually not sure why the ECDHE tests passes - the OpenSSL docs claim that the TLS 1.3 suites aren't in the groups that we're activating.

File: systest/build.rs (hunk 3 above)
Comment (sfackler) [reply systest/build.rs]: Fixed

File: openssl-sys/build.rs (hunk 1 above)
Comment (Ralith) [reply openssl-sys/build.rs]: > Several constants have changed values between 1.1.0 and 1.1.1

That's odd, they seemed pretty clear about that specific type of change being forbidden. Do you recall any specifics?

File: openssl/src/ssl/test.rs (hunk 2 above)
Comment (sfackler) [reply openssl/src/ssl/test.rs]: Oh right, we only run that on 1.0.1 and 1.0.2.

File: openssl-sys/build.rs (hunk 1 above)
Comment (sfackler) [reply openssl-sys/build.rs]: https://github.com/sfackler/rust-openssl/pull/839/files#diff-ceca4680e3c63bdc67652c99c23c0db9

SSL_OP_NO_SSL_MASK changed in 1.1.1, and SSL_OP_ALL changed in 1.1.0f.

**Pull request merged**